clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -3889,24 +4116,196 @@ scheduler_tmpl_t<RecordType, ReaderType>::set_output_active(output_ordinal_t out
         return sched_type_t::STATUS_OK;
-    outputs_[output].active = active;
+    outputs_[output].active->store(active, std::memory_order_release);
     VPRINT(this, 2, "Output stream %d is now %s\n", output,
            active ? "active" : "inactive");
-    std::lock_guard<mutex_dbg_owned> guard(sched_lock_);
+    std::vector<input_ordinal_t> ordinals;
     if (!active) {
         // Make the now-inactive output's input available for other cores.
         // This will reset its quantum too.
         // We aren't switching on a just-read instruction not passed to the consumer,
         // if the queue is empty.
-        if (inputs_[outputs_[output].cur_input].queue.empty())
-            inputs_[outputs_[output].cur_input].switching_pre_instruction = true;
-        set_cur_input(output, INVALID_INPUT_ORDINAL);
+        input_ordinal_t cur_input = outputs_[output].cur_input;
+        if (cur_input != INVALID_INPUT_ORDINAL) {
+            if (inputs_[cur_input].queue.empty())
+                inputs_[cur_input].switching_pre_instruction = true;
+            set_cur_input(output, INVALID_INPUT_ORDINAL);
+        }
+        // Move the runqueue to other outputs.
+        {
+            auto lock = acquire_scoped_output_lock_if_necessary(output);
+            while (!outputs_[output].runqueue.queue.empty()) {
+                input_info_t *tomove = outputs_[output].runqueue.queue.top();
+                ordinals.push_back(tomove->index);
+                outputs_[output].runqueue.queue.pop();
+            }
+        }
     } else {
         outputs_[output].waiting = true;
     }
-    return sched_type_t::STATUS_OK;
+    return rebalance_queues(output, ordinals);
+}
+
+template <typename RecordType, typename ReaderType>
+void
+scheduler_tmpl_t<RecordType, ReaderType>::print_queue_stats()
+{
+    size_t unsched_size = 0;
+    {
+        std::lock_guard<mutex_dbg_owned> unsched_lock(*unscheduled_priority_.lock);
+        unsched_size = unscheduled_priority_.queue.size();
+    }
+    int live = live_input_count_.load(std::memory_order_acquire);
+    VPRINT(this, 1, "inputs: %zd scheduleable, %zd unscheduled, %zd eof\n",
+           live - unsched_size, unsched_size, inputs_.size() - live);
+    for (unsigned int i = 0; i < outputs_.size(); ++i) {
+        auto lock = acquire_scoped_output_lock_if_necessary(i);
+        VPRINT(this, 1, "  out #%d: running #%d; %zd in queue; %d blocked\n", i,
+               // XXX: Reading this is racy; we're ok with that.
+               outputs_[i].cur_input, outputs_[i].runqueue.queue.size(),
+               outputs_[i].runqueue.num_blocked);
+    }
+}
+
+template <typename RecordType, typename ReaderType>
+typename scheduler_tmpl_t<RecordType, ReaderType>::stream_status_t
+scheduler_tmpl_t<RecordType, ReaderType>::rebalance_queues(
+    output_ordinal_t triggering_output, std::vector<input_ordinal_t> inputs_to_add)
+{
+    std::thread::id who;
```

Comment:
I think it might be more descriptive to call this "nobody" or "no_thread" to make it clear the expected value is no valid thread.

Reply:
Done.